cmd/shipperctl/configurator/cluster.go

```diff
@@ -280,3 +294,175 @@ func loadKubeConfig(context, kubeConfigFile string) (*rest.Config, error) {
 
 	return clientConfig.ClientConfig()
 }
+
+func (c *Cluster) CreateCertificateSigningRequest(csr []byte) error {
+	certificateSigningRequest := &certificatesv1beta1.CertificateSigningRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: shipperCSRName,
+		},
+		Spec: certificatesv1beta1.CertificateSigningRequestSpec{
+			Request: csr,
+			Groups:  []string{"system:authenticated"},
+			Usages: []certificatesv1beta1.KeyUsage{
+				certificatesv1beta1.UsageServerAuth,
+				certificatesv1beta1.UsageDigitalSignature,
+				certificatesv1beta1.UsageKeyEncipherment,
+			},
+		},
+	}
+
+	_, err := c.KubeClient.CertificatesV1beta1().CertificateSigningRequests().Create(certificateSigningRequest)
+	return err
+}
+
+func (c *Cluster) ApproveShipperCSR() error {
+	csr, err := c.KubeClient.CertificatesV1beta1().CertificateSigningRequests().Get(shipperCSRName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	approvedCondition := certificatesv1beta1.CertificateSigningRequestCondition{
+		Type:    certificatesv1beta1.CertificateApproved,
+		Reason:  "ShipperctlApprove",
+		Message: "Automatically approved by shipperctl",
+	}
+
+	csr.Status.Conditions = append(csr.Status.Conditions, approvedCondition)
+	_, err = c.KubeClient.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(csr)
+
+	return err
+}
+
+// FetchCertificateFromCSR continually fetches the Shipper CSR until
+// it is populated with a certificate and then returns the PEM-encoded
+// certificate from the Status. This is a blocking function.
+//
+// Note that the returned certificate is already PEM-encoded.
+func (c *Cluster) FetchCertificateFromCSR() ([]byte, error) {
+	for retries := 0; retries < 10; retries++ {
+		csr, err := c.KubeClient.CertificatesV1beta1().CertificateSigningRequests().Get(shipperCSRName, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		if len(csr.Status.Certificate) == 0 {
+			// Pause to give the server some time to sign and populate the certificate
+			time.Sleep(1 * time.Second)
```

Comment:
I think in a long-running process running in the background it would make sense, but since this is a command-line application with no spinners, I'd think that failing fast is better than hanging with no feedback.

What do you think?

Reply:
Not critical indeed. Let's skip.